---
title: "Recombination analyses on the ostrich PAR"
output:
  html_document: default
  pdf_document: default
---

#### by Homa Papoli Yazdi

```{r load-packages, include=FALSE}
rm(list=ls())
library(magrittr)
library(knitr)
library(dplyr)
library(reshape2) 
library(devtools)
library(ggplot2)
library(gridExtra)
library(data.table)
```

## Linkage disequilibrium
```{r}
data_dir = "../results/manuscript_tables"
LD_file = paste(data_dir,"/LD/LD_chromosome_plot/black.Z.coordinate.LD.05-500.200kbBin.50.kbStep.Int.v2.out", sep="")
black_LD <- read.table(LD_file)
colnames(x = black_LD) = c("Chromosome", "Window_start", "Window_end", "Dprime_LD",
                           "LOD_Dprime", "r_squared", "CIlow_rsq", "CIHi_rsq",
                           "mean_SNP_distance", "NUM_snp_pairs")
# LD decay
scaf26 <- read.table(paste(data_dir,"/LD/scaf.split/black.superscaffold26.LDdecay.bin", sep=""))
scaf54<- read.table(paste(data_dir,"/LD/scaf.split/black.superscaffold54.LDdecay.bin", sep=""))
scaf35<- read.table(paste(data_dir,"/LD/scaf.split/black.superscaffold35.LDdecay.bin", sep=""))
scaf36<- read.table(paste(data_dir,"/LD/scaf.split/black.superscaffold36.LDdecay.bin", sep=""))
scaf62<- read.table(paste(data_dir,"/LD/scaf.split/black.superscaffold62.LDdecay.bin", sep=""))
scaf67<- read.table(paste(data_dir,"/LD/scaf.split/black.superscaffold67.LDdecay.bin", sep=""))
scaf69_1<- read.table(paste(data_dir,"/LD/scaf.split/black.superscaffold69-1.LDdecay.bin", sep=""))
scaf93<- read.table(paste(data_dir,"/LD/scaf.split/black.superscaffold93.LDdecay.bin", sep=""))
scaf63<- read.table(paste(data_dir,"/LD/scaf.split/black.superscaffold63.LDdecay.bin", sep=""))
scaf88<- read.table(paste(data_dir,"/LD/scaf.split/black.superscaffold88.LDdecay.bin", sep=""))
scaf83<- read.table(paste(data_dir,"/LD/scaf.split/black.superscaffold83.LDdecay.bin", sep=""))
scaf92<- read.table(paste(data_dir,"/LD/scaf.split/black.superscaffold92.LDdecay.bin", sep=""))
```

## Mean and sd for LD in PAR and nonPAR
```{r}
# LD for PAR and non-PAR
black_LDPAR <- black_LD[black_LD$Window_start < 53065700,]
black_LDnonPAR <- black_LD[black_LD$Window_start > 53065700,]

# Black mean and sd
mean(black_LD$r_squared, na.rm = T)
sd(black_LD$r_squared, na.rm = T)

mean(black_LDPAR$r_squared)
sd(black_LDPAR$r_squared)

mean(black_LDnonPAR$r_squared, na.rm = T)
sd(black_LDnonPAR$r_squared, na.rm = T)

t.test(black_LDPAR$r_squared, black_LDnonPAR$r_squared)
```

## Comparison of LD decay in the PAR and nonPAR
```{r}
plot(scaf26$V1, scaf26$V2)
PAR_ld <- c(scaf26$V2, scaf54$V2, scaf35$V2)
nonPAR_ld <- c(scaf62$V2, scaf63$V2, scaf67$V2, scaf69_1$V2, scaf83$V2, scaf88$V2, scaf92$V2, scaf93$V2)
# mean of LD decay PAR
mean(PAR_ld)
sd(PAR_ld)
# mean of LD decay nonPAR
mean(nonPAR_ld)
sd(nonPAR_ld)
```

## Measuring LD across the PAR-nonPAR boundary 
```{r}




```


## Genetic map
```{r echo=FALSE}
working_dir = "~/Documents/projects/ostrich_Z/results/manuscript_tables"
mb = 10^6
female_map <- read.table(paste(working_dir, "/linkage_map/LGZ3.female.lifted.cleaned.bed.txt", sep = ""))
colnames(female_map) <- c("Chr", "Pos", "PosPlus1", "mapname", "cM", "scaffold")
male_map <- read.table(paste(working_dir, "/linkage_map/LGZ3.male.lifted.cleaned.bed.txt", sep = ""))
colnames(male_map) <- c("Chr", "Pos", "PosPlus1", "mapname", "cM", "scaffold")

sex_averaged_map <- inner_join(female_map, male_map, by = "Pos")
sex_averaged_cM <- (sex_averaged_map$cM.x + sex_averaged_map$cM.y)/2
Pos <- sex_averaged_map$Pos
female_cM <- sex_averaged_map$cM.x
male_cM <- sex_averaged_map$cM.y
sex_averaged_map <- data.frame(Pos, female_cM, male_cM,sex_averaged_cM)
head(sex_averaged_map)
```
## LOESS for smoothing of the three maps
```{r}
female_map_loessMod25 <- loess(cM ~ Pos, data=female_map, span=0.25) # 25% smoothing span
male_map_loessMod25 <- loess(cM ~ Pos, data=male_map, span=0.25) # 25% smoothing span
sex_averaged_map_loessMod25 <- loess(sex_averaged_cM ~ Pos, data=sex_averaged_map, span=0.25) # 25% smoothing span
female_map_smoothed25 <- predict(female_map_loessMod25)
male_map_smoothed25 <- predict(male_map_loessMod25)
sex_average_map_smoothed25 <- predict(sex_averaged_map_loessMod25)
```
## Plotting genetic map 
```{r echo=FALSE}
# Shade the area where PAR-nonPAR boundary is residing
plot(NULL, xlim = c(0, 82), ylim = c(0, 95), ylab = "Genetic map position (cM)", xlab = "Physical position (Mb)")
rect(51738942/mb, -5, 53065777/mb, 100, lty = 2, col = "lightgrey")

# Plot the points for the three genetic maps, female, male and sex-averaged
points(male_map$Pos/mb, male_map$cM, col = "blue")
lines(male_map$Pos/mb, male_map_smoothed25, col = "blue")
points(female_map$Pos/mb, female_map$cM, col = "red")
lines(female_map$Pos/mb, female_map_smoothed25, col = "red")
points(sex_averaged_map$Pos/mb, sex_averaged_map$sex_averaged_cM, col = "darkorchid1")
lines(sex_averaged_map$Pos/mb, sex_average_map_smoothed25, col = "darkorchid1")
legend(x = -2, y = 98, legend = c("male", "female", "sex-averaged"), fill = c("blue", "red", "darkorchid1"), box.lty = 0)
```

## Recombination frequency per Mb

Using interpolated data for males, females and the sex-averaged map, I use the 
Kosambi map function to obtain an estimate of recombination fraction per meiosis.

```{r echo = FALSE}
rate_function <- function(genetic_map, len_markers, column, span, first_marker = 1113306, mb = 10^6){
  start <- genetic_map$Pos[1:len_markers-1]
  end <- genetic_map$Pos[2:len_markers]-1
  pair_cm <- abs(diff(genetic_map[,column]))
  length_region <- end - start
  
  # Calculate recombination frequency for each pair of markers
  kosambi_r_length_region = c()
  c = 0
  for( i in pair_cm){
    c = c + 1
    kosambi_r_length_region[c] = 0.5 * ((exp(4*(i/100))-1)/(exp(4*(i/100))+1))
  }
  pair_cm_per_site <- pair_cm/(end-start)
  kosambi_r_per_site <- kosambi_r_length_region/(end-start)
  df <- data.frame(start, end, pair_cm, pair_cm_per_site, kosambi_r_length_region, kosambi_r_per_site, length_region)
  df1 <- df %>% 
    rowwise() %>% 
    mutate(row = list(seq(from = start, to = end))) %>% 
    ungroup() %>% 
    tidyr::unnest(row) %>% 
    mutate(group = (row - first_marker) %/% mb)
  
  df2 <- df1 %>% 
    group_by(group) %>% 
    summarise(
      start = min(row), 
      end = max(row), 
      pair_cm_per_site = mean(pair_cm_per_site)*mb, 
      .groups = "drop"
    ) %>% 
    select(-group)
  rate_loess <- loess(pair_cm_per_site ~ start, data=df2, span=span, na.action = na.exclude)
  rate_smoothed <- predict(rate_loess)
  return(list(df, df2, rate_loess, rate_smoothed))
}
```

## Recombination frequency
```{r echo = FALSE}
female_rate <- rate_function(genetic_map = sex_averaged_map, len_markers = 193, column = 2, span = 0.3)
male_rate <- rate_function(genetic_map = sex_averaged_map, len_markers = 193, column = 3, span = 0.3)
sex_averaged_rate <- rate_function(genetic_map = sex_averaged_map, len_markers = 193, column = 4, span = 0.3)

head(female_rate[[1]])
head(male_rate[[1]])
head(sex_averaged_rate[[1]])

plot(NULL, xlim = c(0, 82), ylim = c(0,7), ylab = "cM/Mb", xlab = "Physical position (Mb)")
rect(51738942/mb, -5, 53065777/mb, 100, lty = 2, col = "lightgrey")

points((female_rate[[2]]$start+female_rate[[2]]$end)/(2*mb), female_rate[[2]]$pair_cm_per_site, col = "red", pch = 20)
lines((female_rate[[2]]$start+female_rate[[2]]$end)/(2*mb), female_rate[[4]], col = "red")

points((male_rate[[2]]$start+male_rate[[2]]$end)/(2*mb), male_rate[[2]]$pair_cm_per_site, col = "blue", pch = 20)
lines((male_rate[[2]]$start+male_rate[[2]]$end)/(2*mb), male_rate[[4]], col = "blue")

points((sex_averaged_rate[[2]]$start+sex_averaged_rate[[2]]$end)/(2*mb), sex_averaged_rate[[2]]$pair_cm_per_site, col = "darkorchid1", pch = 20)
lines((sex_averaged_rate[[2]]$start+sex_averaged_rate[[2]]$end)/(2*mb), sex_averaged_rate[[4]], col = "darkorchid1")
```

## Genetic length in PAR in females, males and sex-averaged where PAR is defined at 53065777
```{r echo=FALSE}
female_PAR_length = 81.163  
male_PAR_length = 44.199
sex_averaged_PAR_length = 62.6810

female_PAR_r = 0.5 * ((exp(4*(female_PAR_length/100))-1)/(exp(4*(female_PAR_length/100))+1))
male_PAR_r = 0.5 * ((exp(4*(male_PAR_length/100))-1)/(exp(4*(male_PAR_length/100))+1))
sex_averaged_PAR_r = 0.5 * ((exp(4*(sex_averaged_PAR_length/100))-1)/(exp(4*(sex_averaged_PAR_length/100))+1))
```

### Recombination frequency female PAR (81.163 cM)
```{r echo=FALSE}
print(female_PAR_r)
```
### Recombination frequency male PAR (44.199 cM)
```{r echo=FALSE}
print(male_PAR_r)
```
### Recombination frequency sex averaged PAR (62.6810 cM)
```{r echo=FALSE}
print(sex_averaged_PAR_r)
```

## Population scaled recombination rate
We have two file types: A rates file and a position file
The rates file looks as follows

5000 2000

135.225 0.001319 0.000273 0.000273 .... 

171.977 0.004708 0.000273 0.000338 ....

5000 is the number of samples from running MCMC for 10 million times and sampling
every 2000 times
2000 is the number of markers in a window
The first number in every row (135.225 and 171.977) gives the rho for the window
of 2000 SNPs
Each row is a MCMC run, therefore we should have 5001 rows 
Every column except the first is the per site rho for the two consecutive markers,
therefore we should have 2000 columns (2000 SNPs give 1999 rho estimates)
Since I want to have contiguous rhos, I remove the last 249 values from the first
window and the first 249 values from the second window for each two consecutive windows.
To obtain the rho for those 1750 markers, I multiply each estimate rho per site by 
the length of the two consecutive markers. 

The function to summarise the interval output of LDhat is called windows_together(). 
Function returns rowSums(g) which is rho for the window consisting of the 
chosen number of consecutive markers for each MCMC run after having removed the
burn-in run. Function returns means: This is the mean per site rho for each consecutive two markers across
all MCMC runs after having removed the burn-in, finally, pos: position of each consecutive marker in a window
```{r echo=FALSE}
windows_together <- function(rates.file, locs.file, start = 1, end = 1750, burn.in=30, nsamples = 5000, nsnps = 2000) {
  rates <- read.table(rates.file, skip = 1, fill = T)
  low = as.integer(nrow(rates)*burn.in/100)
  df <- data.frame(rates)[low:nsamples,2:nsnps]
  distances <- read.table(locs.file, skip = 1)  
  g <- data.frame(mapply(`*`,df[,start:end],diff(distances$V1)[start:end]))
  
  pos <- as.vector(as.matrix(distances))[start:end+1]
  means <-apply(df[low:nrow(df),], 2, mean, na.rm=T)[start:end]
  return(list(rowSums(g), means, pos))
}
```


```{r eval=FALSE, echo=FALSE}
# Rho for superscaffold26
superscaffold26_map_length = rep(0,101)
first_scaffold_26 <- windows_together(rates.file = "superscaffold26.2000.500.1.rates.txt", locs.file ="superscaffold26.2000.500.1.pos.txt", start = 1, end = 1750)
superscaffold26_map_length[1] = mean(first_scaffold_26[[1]])
first_mean_26 = first_scaffold_26[[2]]
first_pos_26 = first_scaffold_26[[3]]
last_scaffold_26 <- windows_together(rates.file = "superscaffold26.2000.500.101.rates.txt", locs.file ="superscaffold26.2000.500.101.pos.txt", start = 249, end = 833, nsnps = 834)
last_mean_26 = last_scaffold_26[[2]]
last_pos_26 = last_scaffold_26[[3]]

r26 <- c(first_mean_26) # mean per site rho for each consecutive two markers. this contains 1750 values.
p26 <- c(first_pos_26) # position of each consecutive marker in a window. this contains 1750 values.
for(i in seq(2,100)){
  rate.name = paste("superscaffold26.2000.500.", i, ".rates.txt", sep = "")
  locs.name = paste("superscaffold26.2000.500.", i, ".pos.txt", sep = "")
  print(rate.name)
  print(locs.name)
  mapl = windows_together(rates.file = rate.name, locs.file = locs.name, start = 249, end = 1750)
  superscaffold26_map_length[i] = mean(mapl[[1]])
  
  means = mapl[[2]]
  pos = mapl[[3]]
  r26 <- c(r26, means)
  p26 <- c(p26, pos)  
}

r26 <- c(r26, last_mean_26) # This will give us a vector containing all the pairwise rhos per site for each consecutive two markers
# across all positions.
p26 <- c(p26, last_pos_26)

superscaffold26_map_length[101] = mean(last_scaffold_26[[1]])
s26_map_length <- sum(superscaffold26_map_length)

# Rho for superscaffold54
superscaffold54_map_length = rep(0,59)
first_scaffold_54 <- windows_together(rates.file = "superscaffold54.2000.500.1.rates.txt", locs.file ="superscaffold54.2000.500.1.pos.txt", start = 1, end = 1750)
superscaffold54_map_length[1] = mean(first_scaffold_54[[1]])
first_mean_54 = first_scaffold_54[[2]]
first_pos_54 = first_scaffold_54[[3]]
last_scaffold_54 <- windows_together(rates.file = "superscaffold54.2000.500.59.rates.txt", locs.file ="superscaffold54.2000.500.59.pos.txt", start = 249, end = 1475, nsnps = 1476)
last_mean_54 = last_scaffold_54[[2]]
last_pos_54 = last_scaffold_54[[3]]

r54 <- c(first_mean_54)
p54 <- c(first_pos_54)
for(i in seq(2,58)){
  rate.name = paste("superscaffold54.2000.500.", i, ".rates.txt", sep = "")
  locs.name = paste("superscaffold54.2000.500.", i, ".pos.txt", sep = "")
  print(rate.name)
  print(locs.name)
  mapl = windows_together(rates.file = rate.name, locs.file = locs.name, start = 249, end = 1750)
  superscaffold54_map_length[i] = mean(mapl[[1]])
  
  means = mapl[[2]]
  pos = mapl[[3]]
  r54 <- c(r54, means)
  p54 <- c(p54, pos)  
}

r54 <- c(r54, last_mean_54)
p54 <- c(p54, last_pos_54)

superscaffold54_map_length[59] = mean(last_scaffold_54[[1]])
s54_map_length <- sum(superscaffold54_map_length)

# Rho for superscaffold35
superscaffold35_map_length = rep(0,19)
first_scaffold_35 <- windows_together(rates.file = "superscaffold35.2000.500.1.rates.txt", locs.file ="superscaffold35.2000.500.1.pos.txt", start = 1, end = 1750)
superscaffold35_map_length[1] = mean(first_scaffold_35[[1]])
first_mean_35 = first_scaffold_35[[2]]
first_pos_35 = first_scaffold_35[[3]]
last_scaffold_35 <- windows_together(rates.file = "superscaffold35.2000.500.19.rates.txt", locs.file ="superscaffold35.2000.500.19.pos.txt", start = 249, end = 1319, nsnps = 1320)
last_mean_35 = last_scaffold_35[[2]]
last_pos_35 = last_scaffold_35[[3]]

r35 <- c(first_mean_35)
p35 <- c(first_pos_35)
for(i in seq(2,18)){
  rate.name = paste("superscaffold35.2000.500.", i, ".rates.txt", sep = "")
  locs.name = paste("superscaffold35.2000.500.", i, ".pos.txt", sep = "")
  print(rate.name)
  print(locs.name)
  mapl = windows_together(rates.file = rate.name, locs.file = locs.name, start = 249, end = 1750)
  superscaffold35_map_length[i] = mean(mapl[[1]])
  
  means = mapl[[2]]
  pos = mapl[[3]]
  r35 <- c(r35, means)
  p35 <- c(p35, pos)  
}

r35 <- c(r35, last_mean_35)
p35 <- c(p35, last_pos_35)

superscaffold35_map_length[19] = mean(last_scaffold_35[[1]])
s35_map_length <- sum(superscaffold35_map_length)

# Rho for superscaffold36
superscaffold36_map_length = rep(0,25)
first_scaffold_36 <- windows_together(rates.file = "superscaffold36.2000.500.1.rates.txt", locs.file ="superscaffold36.2000.500.1.pos.txt", start = 1, end = 1750)
superscaffold36_map_length[1] = mean(first_scaffold_36[[1]])
first_mean_36 = first_scaffold_36[[2]]
first_pos_36 = first_scaffold_36[[3]]
last_scaffold_36 <- windows_together(rates.file = "superscaffold36.2000.500.24.rates.txt", locs.file ="superscaffold36.2000.500.24.pos.txt", start = 249, end = 665, nsnps = 666)
last_mean_36 = last_scaffold_36[[2]]
last_pos_36 = last_scaffold_36[[3]]

r36 <- c(first_mean_36)
p36 <- c(first_pos_36)
for(i in seq(2,24)){
  rate.name = paste("superscaffold36.2000.500.", i, ".rates.txt", sep = "")
  locs.name = paste("superscaffold36.2000.500.", i, ".pos.txt", sep = "")
  print(rate.name)
  print(locs.name)
  mapl = windows_together(rates.file = rate.name, locs.file = locs.name, start = 249, end = 1750)
  superscaffold36_map_length[i] = mean(mapl[[1]])
  
  means = mapl[[2]]
  pos = mapl[[3]]
  r36 <- c(r36, means)
  p36 <- c(p36, pos)  
}

r36 <- c(r36, last_mean_36)
p36 <- c(p36, last_pos_36)

superscaffold36_map_length[25] = mean(last_scaffold_36[[1]])
s36_map_length <- sum(superscaffold36_map_length)
```

## Rho per site and cumulative rho for each superscaffold
```{r echo= FALSE}
# Read rho intervals into R
working_dir = "~/Documents/projects/ostrich_Z/results/manuscript_tables"
rho_26 <- read.table(paste(working_dir, "/ldhat_postR_tables/rho.sc26.txt", sep = ""), header = T)
rho_54 <- read.table(paste(working_dir, "/ldhat_postR_tables/rho.sc54.txt", sep = ""), header = T)
rho_35 <- read.table(paste(working_dir, "/ldhat_postR_tables/rho.sc35.txt", sep = ""), header = T)
rho_36 <- read.table(paste(working_dir, "/ldhat_postR_tables/rho.sc36.txt", sep = ""), header = T)
```

## Rho for superscaffold26
```{r echo= FALSE}
plot(rho_26$p26[1:(length(rho_26$p26)-1)], y=rho_26$r26[2:length(rho_26$r26)], type="s", col=rgb(0,0,0.5), 	
     xlab="Position", ylab="Posterior mean rate", main="Posterior mean rates")
plot(rho_26$p26[1:(length(rho_26$p26)-1)]/1000000,cumsum(diff(rho_26$p26)*rho_26$r26[2:length(rho_26$r26)]), type = "l", xlab = "Position", ylab = "4Ner")
```

## Rho for superscaffold54
```{r echo= FALSE}
plot(rho_54$p54[1:(length(rho_54$p54)-1)], y=rho_54$r54[2:length(rho_54$r54)], type="s", col=rgb(0,0,0.5), 	
     xlab="Position", ylab="Posterior mean rate", main="Posterior mean rates")
plot(rho_54$p54[1:(length(rho_54$p54)-1)]/1000000,cumsum(diff(rho_54$p54)*rho_54$r54[2:length(rho_54$r54)]), type = "l", xlab = "Position", ylab = "4Ner")
```

## Rho for superscaffold35
```{r echo= FALSE}
plot(rho_35$p35[1:(length(rho_35$p35)-1)], y=rho_35$r35[2:length(rho_35$r35)], type="s", col=rgb(0,0,0.5), 	
     xlab="Position", ylab="Posterior mean rate", main="Posterior mean rates")
plot(rho_35$p35[1:(length(rho_35$p35)-1)]/1000000,cumsum(diff(rho_35$p35)*rho_35$r35[2:length(rho_35$r35)]), type = "l", xlab = "Position", ylab = "4Ner")
```

## Rho for superscaffold36
```{r echo= FALSE}
plot(rho_36$p36[1:(length(rho_36$p36)-1)], y=rho_36$r36[2:length(rho_36$r36)], type="s", col=rgb(0,0,0.5), 	
     xlab="Position", ylab="Posterior mean rate", main="Posterior mean rates")
plot(rho_36$p36[1:(length(rho_36$p36)-1)]/1000000,cumsum(diff(rho_36$p36)*rho_36$r36[2:length(rho_36$r36)]), type = "l", xlab = "Position", ylab = "4Ner")
```

## Total map length in 4Ner for the PAR
```{r eval = FALSE}
total_PAR_map_length = s36_map_length + s35_map_length + s54_map_length + s26_map_length
print(total_PAR_map_length)
```

## Cumulative rho across PAR from the distal end to the left to the PAR boundary to the right
```{r echo=FALSE}
cum_positions <- c(rho_26$p26[1:(length(rho_26$p26)-1)] , (tail(rho_26$p26, n=1) + rho_54$p54[1:(length(rho_54$p54)-1)]) , (41243377 + rho_35$p35[1:(length(rho_35$p35)-1)]) , (45714693 + rho_36$p36[1:(length(rho_36$p36)-1)]-3869516))

s26_rates <- cumsum(diff(rho_26$p26)*rho_26$r26[2:length(rho_26$r26)])  
s54_rates <- tail(s26_rates, n = 1) + cumsum(rev(diff(rho_54$p54))*rev(rho_54$r54[2:length(rho_54$r54)]))
s35_rates <- tail(s54_rates, n = 1) + cumsum(diff(rho_35$p35)*rho_35$r35[2:length(rho_35$r35)])
s36_rates <- tail(s35_rates, n = 1) + cumsum(rev(diff(rho_36$p36))*rev(rho_36$r36[2:length(rho_36$r36)]))  

cum_rates <- c(s26_rates, s54_rates, s35_rates, s36_rates)
compiled_rates = c(rho_26$r26[2:length(rho_26$r26)], rev(rho_54$r54[2:length(rho_54$r54)]), rho_35$r35[2:length(rho_35$r35)], rev(rho_36$r36[2:length(rho_36$r36)]))

plot(cum_positions[1:(length(cum_positions)-1)]/1000000, y=compiled_rates[2:length(compiled_rates)], type="s", col=rgb(0,0,0.5), 	
     xlab="Position (Mb)", ylab="Posterior mean rate")

plot(cum_positions[1:(length(cum_positions )-1)]/1000000, y=cum_rates[2:length(cum_rates)], type = "l", col=rgb(0,0,0.5), 
     xlab = "Position (Mb)", ylab = "4Ner")
```

cum_positions <- c(rho_26$p26[1:(length(rho_26$p26)-1)] , (tail(rho_26$p26, n=1) + rho_54$p54[1:(length(rho_54$p54)-1)]) , (41243377 + rho_35$p35[1:(length(rho_35$p35)-1)]) , (45714693 + rho_36$p36[1:(length(rho_36$p36)-1)]-3869516))

cum_positions <- c(rho_26$p26[1:(length(rho_26$p26)-1)] , (tail(rho_26$p26, n=1) + rho_54$p54[1:(length(rho_54$p54)-1)]) , (41243377 + rho_35$p35[1:(length(rho_35$p35)-1)]) , (45714693 + rho_36$p36[1:(length(rho_36$p36)-1)]-3869516))

s26_rates <- cumsum(diff(rho_26$p26)*rho_26$r26[2:length(rho_26$r26)])  
s54_rates <- tail(s26_rates, n = 1) + cumsum(rev(diff(rho_54$p54))*rev(rho_54$r54[2:length(rho_54$r54)]))
s35_rates <- tail(s54_rates, n = 1) + cumsum(diff(rho_35$p35)*rho_35$r35[2:length(rho_35$r35)])
s36_rates <- tail(s35_rates, n = 1) + cumsum(rev(diff(rho_36$p36))*rev(rho_36$r36[2:length(rho_36$r36)]))  

s36_rates <- cumsum(rev(diff(rho_36$p36))*rev(rho_36$r36[2:length(rho_36$r36)]))  
s35_rates <- tail(s36_rates, n = 1) + cumsum(diff(rho_35$p35)*rho_35$r35[2:length(rho_35$r35)])
s54_rates <- tail(s35_rates, n = 1) + cumsum(rev(diff(rho_54$p54))*rev(rho_54$r54[2:length(rho_54$r54)]))
s26_rates <- cumsum(diff(rho_26$p26)*rho_26$r26[2:length(rho_26$r26)])  

s26_rates <- cumsum(diff(rho_26$p26)*rho_26$r26[2:length(rho_26$r26)])  
s54_rates <- tail(s26_rates, n = 1) + cumsum(rev(diff(rho_54$p54))*rev(rho_54$r54[2:length(rho_54$r54)]))


cum_rates <- c(s26_rates, s54_rates, s35_rates, s36_rates)


## Calculate window based rho for each r using the sex-averaged smoothed r
```{r}
head(sex_averaged_rate[[2]])
head(sex_averaged_rate[[4]])

start <- sex_averaged_rate[[2]]$start
end <- sex_averaged_rate[[2]]$end
smoothed_rate <- sex_averaged_rate[[4]]
real_rate <- sex_averaged_rate[[2]]$pair_cm_per_site
g <- data.frame(start, end, smoothed_rate, real_rate)

Ne = c()
Ne_real_r = c()
rho = c()
for(i in seq(1,23)){
  w1 <- rho_26[which(rho_26$p26>g[i,1] & rho_26$p26<g[i,2]),]
  w1_sum = sum(diff(w1$p26)*w1$r26[2:length(w1$r26)])
  ne_first_window = w1_sum/(4*g[i,3]*0.01)
  Ne[i] = ne_first_window
  rho[i] = w1_sum
  # Use real r
  if(g[i,4] >= 0.005){
    ne_with_real_r = w1_sum/(4*g[i,4]*0.01)
    Ne_real_r[i] = ne_with_real_r
  }
  else{
    Ne_real_r[i] = NA
  }
}

for(i in seq(24,40)){
  w1 <- rho_54[which(rho_54$p54>g[i,1]-24113305 & rho_54$p54<g[i,2]-24113305),]
  w1_sum = sum(diff(w1$p54)*w1$r54[2:length(w1$r54)])
  ne_first_window = w1_sum/(4*g[i,3]*0.01)
  Ne[i] = ne_first_window
  rho[i] = w1_sum
  # Use real r
  if(g[i,4] >= 0.005){
    ne_with_real_r = w1_sum/(4*g[i,4]*0.01)
    Ne_real_r[i] = ne_with_real_r
  }
  else{
    Ne_real_r[i] = NA
  }
}

for(i in seq(41,45)){
  w1 <- rho_35[which(rho_35$p35>g[i,1]-41588307 & rho_35$p35<g[i,2]-41588307),]
  w1_sum = sum(diff(w1$p35)*w1$r35[2:length(w1$r35)])
  ne_first_window = w1_sum/(4*g[i,3]*0.01)
  Ne[i] = ne_first_window
  rho[i] = w1_sum
  # Use real r
  if(g[i,4] >= 0.005){
    ne_with_real_r = w1_sum/(4*g[i,4]*0.01)
    Ne_real_r[i] = ne_with_real_r
  }
  else{
    Ne_real_r[i] = NA
  }
}

for(i in seq(46,51)){
  w1 <- rho_36[which(rho_36$p36>g[i,1]-46267993+3869516 & rho_36$p36<g[i,2]-46267993+3869516),]
  w1_sum = sum(diff(w1$p36)*w1$r36[2:length(w1$r36)])
  ne_first_window = w1_sum/(4*g[i,3]*0.01)
  Ne[i] = ne_first_window
  rho[i] = w1_sum
  # Use real r
  if(g[i,4] >= 0.005){
    ne_with_real_r = w1_sum/(4*g[i,4]*0.01)
    Ne_real_r[i] = ne_with_real_r
  }
  else{
    Ne_real_r[i] = NA
  }
}

plot(g$start[1:51]/10^6, Ne[1:51], ylab = "Ne", xlab = "Position (Mb)", pch = 20)
plot(g$start[1:51]/10^6, Ne_real_r[1:51], ylab = "Ne", xlab = "Position (Mb)", pch = 20)
```

## Reverse the function
```{r}
rev_pos <- rev(abs(g$start[1:51] - g$start[51]))
rev_Ne <- rev(Ne[1:51])
plot(rev_pos/10^6, rev_Ne, col = "red", pch = 20, ylab = "Ne", xlab = "Position (Mb)")
```

## Real cM/Mb vs. smoothed cM/Mb 
```{r}
plot(sex_averaged_rate[[2]]$start/mb, sex_averaged_rate[[2]]$pair_cm_per_site)
points(sex_averaged_rate[[2]]$start/mb, sex_averaged_rate[[4]], col = "red", pch = 20)
```

## Where is the PAR-nonPAR boundary?
The coordinate of the PAR boundary is located in superscaffold 36 between 3516672 and 3524264. Between these two coordinates, 
there is a gap with length 7592 nucleotides. In scaffold36, anything larger than 3524264 is located on the PAR. In the genetic map, 
this is between the following two rows:
137 chrZ 51738942 51738943 female_input-Z 80.628    superscaffold36:3985614
138 chrZ 53065777 53065778 female_input-Z 81.163    superscaffold36:2658779

In the genetic marker data, it should be between marker superscaffold36:2658779 (in the nonPAR) and marker superscaffold36:3985614 (in the PAR). As of the cumulative position, it should be between chrZ:51738942 (in the PAR) and chrZ:53065777 (in the nonPAR).

# Final table
```{r}
start <- g$start[1:51]
end <- g$end[1:51]
real_rate_map_cM_Mb <- sex_averaged_rate[[2]]$pair_cm_per_site[1:51]
smooth_rate_map_cM_Mb <- g$smoothed_rate[1:51]
rho <- rho
Ne_smooth_r <- Ne
Ne_real_r <- Ne_real_r[1:51]

rho_r_Ne_table <- data.frame(start, end, real_rate_map_cM_Mb, smooth_rate_map_cM_Mb, rho, Ne_smooth_r, Ne_real_r)
head(rho_r_Ne_table)
write.table(rho_r_Ne_table, "~/Documents/projects/ostrich_Z/results/manuscript_tables/rho_r_Ne_table.txt", quote = FALSE, sep = "\t", row.names = FALSE, col.names = TRUE)
```
## Plot of smoothed vs. real values of r
```{r}

plot(rho_r_Ne_table$start/(10^6), rho_r_Ne_table$rho, pch = 20, 
     xlab = "Position (Mb)", ylab = "Rho")

plot(rho_r_Ne_table$real_rate_map_cM_Mb, rho_r_Ne_table$smooth_rate_map_cM_Mb, pch = 20, 
     xlab = "Real rate in 1 Mb windows", ylab = "Smoothed rate in 1 Mb windows")

plot(rho_r_Ne_table$Ne_real_r, rho_r_Ne_table$Ne_smooth_r, pch = 20, 
     xlab = "Ne (real r) in 1 Mb windows", ylab = "Ne (smoothed r) in 1 Mb windows")

plot(rho_r_Ne_table$real_rate_map_cM_Mb, rho_r_Ne_table$Ne_real_r, pch = 20, 
     xlab = "Real rate in 1 Mb windows", ylab = "Ne (real r) in 1 Mb windows")

plot(rho_r_Ne_table$smooth_rate_map_cM_Mb, rho_r_Ne_table$Ne_smooth_r, pch = 20, 
     xlab = "Smoothed rate in 1 Mb windows", ylab = "Ne (smoothed r) in 1 Mb windows")

plot(rho_r_Ne_table$rho, rho_r_Ne_table$Ne_real_r, pch = 20, 
     xlab = "rho in 1 Mb windows", ylab = "Ne (real r) in 1 Mb windows")

plot(rho_r_Ne_table$rho, rho_r_Ne_table$Ne_smooth_r, pch = 20, 
     xlab = "rho in 1 Mb windows", ylab = "Ne (smoothed r) in 1 Mb windows")
```

## Final table reversed 
```{r}
start_rev <- rev(abs(g$start[1:51] - g$start[51]))
end_rev <- rev(abs(g$end[1:51] - g$end[51])+10^6)
real_rate_map_cM_Mb_rev <- rev(sex_averaged_rate[[2]]$pair_cm_per_site[1:51])
smooth_rate_map_cM_Mb_rev <- rev(g$smoothed_rate[1:51])
rho_rev <- rev(rho)
Ne_smooth_r_rev <- rev(Ne)
Ne_real_r_rev <- rev(Ne_real_r[1:51])

rho_r_Ne_table_rev <- data.frame(start_rev, end_rev, real_rate_map_cM_Mb_rev, smooth_rate_map_cM_Mb_rev, rho_rev, Ne_smooth_r_rev, Ne_real_r_rev)

head(rho_r_Ne_table_rev)

plot(rho_r_Ne_table_rev$start_rev/10^6, rho_r_Ne_table_rev$rho, xlab = "Position (Mb)", ylab = "4Ner - Red is reverse")
points(rho_r_Ne_table$start/10^6, rho_r_Ne_table$rho, col = "red", pch = 20)

plot(rho_r_Ne_table_rev$start_rev/10^6, rho_r_Ne_table_rev$Ne_smooth_r_rev, xlab = "Position (Mb)", ylab = "Ne - Red is reverse")
points(rho_r_Ne_table$start/10^6, rho_r_Ne_table$Ne_smooth_r, col = "red", pch = 20)

write.table(rho_r_Ne_table_rev, "~/Documents/projects/ostrich_Z/results/manuscript_tables/rho_r_Ne_table_reversed.txt", quote = FALSE, sep = "\t", row.names = FALSE, col.names = TRUE)
```

## Table for superscaffold36 close to the boundary
```{r}
start_sc36 <- rho_36$p36[1:10000][1:(length(rho_36$p36[1:10000])-1)]
end_sc36 <- rho_36$p36[1:10000][2:length(rho_36$p36[1:10000])]
rho_sc36 <- diff(rho_36$p36[1:10000])*rho_36$r36[1:10000][2:length(rho_36$r36[1:10000])]
rho_scaffold_36 <- data.frame(start_sc36, end_sc36, rho_sc36)

chr_start <- abs((rho_scaffold_36$start_sc36-3525136)-5260899) + 46315381
chr_end <- abs((rho_scaffold_36$end_sc36-3525136)-5260899) + 46315381

chr_start <- abs((rho_scaffold_36$start_sc36-3525136)-5869039) + 46315381
chr_end <- abs((rho_scaffold_36$end_sc36-3525136)-5869039) + 46315381

rho_scaffold_36 <- data.frame(start_sc36, end_sc36, chr_start, chr_end, rho_sc36)

plot(rho_scaffold_36$start_sc36, cumsum(rho_scaffold_36$rho_sc36), ylab = "4Ner", xlab = "Position bp")
plot(rho_scaffold_36$chr_start, cumsum(rho_scaffold_36$rho_sc36), ylab = "4Ner", xlab = "Position bp")

geneticmap <- read.table("~/Documents/projects/ostrich_Z/results/manuscript_tables/rho_r_Ne_table.txt", header=T)

head(geneticmap)

real_rate_map_cM_Mb = c()
smooth_rate_map_cM_Mb = c()
Ne_smooth_r = c()
Ne_real_r = c()
k = 0
for(i in rho_scaffold_36$chr_start) {
k = k + 1
c = 0
  for(j in geneticmap$start){
  c = c + 1
  if(i >= j && i <= j + 1000000){
  real_rate_map_cM_Mb[k] = geneticmap[c,3]
  smooth_rate_map_cM_Mb[k] = geneticmap[c,4]
  Ne_smooth_r[k] = geneticmap[c,6]
  Ne_real_r[k] = geneticmap[c,7]
  }
  }
}

rho_scaffold_36 <- data.frame(start_sc36, end_sc36, chr_start, chr_end, rho_sc36, real_rate_map_cM_Mb, smooth_rate_map_cM_Mb, Ne_smooth_r, Ne_real_r)

write.table(rho_scaffold_36, "~/Documents/projects/ostrich_Z/results/manuscript_tables/rho_scaffold_36.txt", quote = FALSE, sep = "\t", row.names = FALSE, col.names = TRUE)

```




















